### DISCUSSION TITLE
Multicast Addresses

### DISCUSSION THREAD

**Discussion by hummerbaendiger** [Iam1337/extOSC#27] [Q&A]
Hello,

Is there a way to enable the communication to run on multicast IP addresses? I haven't looked too deep into the implementation yet so there is a good chance I've overlooked it... 😊 

The background why I'm asking is that I used OSC communication quite a lot when I was working with Ventuz (3D realtime rendering engine for presentations and broadcast). 
Using a multicast IP address (e.g. 230.2.2.2) to send messages was always very convenient since you don't need to know the exact IP of a receiver machine. All receivers that listen to this address would receive the messages simultaneously.
Also setting up a universal remote (e.g. as a phone app) is quite convenient as you just need to be in the same network and know a maintenance multicast address to send messages to.

**Comment by hummerbaendiger** [comment]
I found a way to do a quick and dirty implementation but it is probably dodgy as hell:
When adding a few lines to the Create() method in OSCStandaloneManager, the UdpClient can be set to receiving on multicast addresses. Though it is probably only working for IPv4 and doesn't do any safety checks 😅 

```
public static UdpClient Create(string localHost, int localPort)
{
	IPAddress address = IPAddress.Parse(localHost);
	byte[] bytes = address.GetAddressBytes();
	bool isMultiCast = bytes[0] > 224 && bytes[0] < 240;

	var localEndPoint = new IPEndPoint(IPAddress.Parse(localHost), localPort);
	var clientInfo = _clientsList.FirstOrDefault(c => c.Client.Client.LocalEndPoint != null &&
															  c.Client.Client.LocalEndPoint.Equals(localEndPoint));

	if (clientInfo == null)
	{
		clientInfo = new ClientInfo();
		if (!isMultiCast)
		{
			clientInfo.Client = new UdpClient(localEndPoint);
		}
		else
		{
			clientInfo.Client = new UdpClient(localPort, AddressFamily.InterNetwork);
			clientInfo.Client.JoinMulticastGroup(address);
		}
				
		clientInfo.Client.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
		clientInfo.Client.DontFragment = true;

		_clientsList.Add(clientInfo);
	}

	clientInfo.Links++;

	return clientInfo.Client;
}
```

The receiver must then be set to LocalHostMode=Custom and get the common multicast address as Local Host address.

**Reply by Thebinoman** [reply]
Awesome!
What about sending multicast messages? Can you implement that as well?

**Comment by Iam1337** [comment]
Hello! Never worked with Multicast, your solution looks quite interesting. I'll look at this point in more detail, maybe it will be possible to make support for Multicast.
